File: 2019prep/2018_d02.py
from collections import Counter
from itertools import combinations

from aocd.models import Puzzle


def part1(boxes):
    """
    >>> part1(["abcdef", "bababc", "abbcde", "abcccd", "aabcdd", "abcdee", "ababab"])
    12
    """
    twos = 0
    threes = 0
    for box in boxes:
        counts = Counter(box).values()
        if 2 in counts:
            twos += 1
        if 3 in counts:
            threes += 1
    return twos * threes


def part2(boxes):
    """
    >>> part2(["abcde", "fghij", "klmno", "pqrst", "fguij", "axcye", "wvxyz"])
    'fgij'
    """
    for box1, box2 in combinations(boxes, 2):
        shared = [char1 for char1, char2 in zip(box1, box2) if char1 == char2]
        if len(shared) == len(box1) - 1:
            return ''.join(shared)


if __name__ == "__main__":
    import doctest
    doctest.testmod()

    puzzle = Puzzle(year=2018, day=2)
    boxes = puzzle.input_data.split('\n')
    puzzle.answer_a = part1(boxes)
    puzzle.answer_b = part2(boxes)

File: 2019prep/2018_d02.exs
defmodule Day1 do
  def counts(box) do
    counts = box
    |> String.graphemes
    |> Enum.sort
    |> Enum.group_by(fn(char) -> char end)
    |> Map.values
    |> Enum.map(&Enum.count/1)
    [Enum.member?(counts, 2), Enum.member?(counts, 3)]
  end

  def part1(boxes) do
    counts = boxes |> Enum.map(&Day1.counts/1)
    twos = counts |> Enum.filter(fn ([two, _]) -> two end) |> Enum.count
    threes = counts |> Enum.filter(fn ([_, three]) -> three end) |> Enum.count
    twos * threes
  end

  def part2(boxes) do
  end
end

ExUnit.start

defmodule Day1Test do
  use ExUnit.Case

  test "part1" do
    boxes = [
      "abcdef",
      "bababc",
      "abbcde",
      "abcccd",
      "aabcdd",
      "abcdee",
      "ababab",
    ]
    assert boxes |> Day1.part1() == 12
  end

  test "part2" do

  end
end

input = with {:ok, raw} = AdventOfCodeHelper.get_input(2018, 2), do: raw
|> String.trim()
|> String.split("\n")

input
|> Day1.part1
|> IO.puts()

input
|> Day1.part2
|> IO.puts()
